### DISCUSSION TITLE
Implementing a single-worker queue module in a load-balanced environment

### DISCUSSION THREAD

**Discussion by aaronbushnell** [craftcms/cms#12337] [Q&A] [answered]
This is related to #11507 and _some_ of #9905.

We have a couple load-balanced Craft sites that have `runQueueAutomatically` set to `false`. Because this is `false` we send the `queue/run` command to one of the web servers in the load-balanced pool.

Here's how this process works:
1. Grab a random web server from the load-balancer pool
2. Run `queue/run`

Due to our hosting setup step 1 won't always grab the _same_ web server. It just grabs a single server from the pool. That means the next time this process runs it could introduce a second queue worker (if it hits a different server).

Ideally we'd setup a single server responsible for the queue, but we're unable to do that based on the current infrastructure. Instead, I was hoping to solve this with a bit of code. Here's what I was thinking:

1. Run a _new_ process called `proxy-queue/run`
2. This process checks if the queue is already running
3. If it's not running, fire `Craft::$app->getQueue()->run()`
4. If it is running, just exit

Like so:
```php
class Queue extends Controller
{
    public function actionRun()
    {
        // Exit if there are running jobs
        if (Craft::$app->getQueue()->getHasReservedJobs()) {
            ExitCode::OK;
        }

        // Otherwise run the queue as normal
        Craft::$app->getQueue()->run();

        ExitCode::OK;
    }
}
```

Would this solution work as a workaround to our infrastructure setup? Or is there something I'm missing?

**Comment by brandonkelly** [comment, marked as answer]
`getHasReservedJobs()` can be a little unreliable if a job resulted in a fatal error (e.g. a resource limit was exceeded).

Instead I’ve added support for a new `--isolated` option to **all** CLI commands for Craft 4.4 (#12350), which can be passed to ensure the command is only run once at a time. Any duplicate runs will be aborted with `The <command name> command is already running.` output to stderr and exit code `1`.

Couple examples:

```sh
php craft up --isolated
```

```sh
php craft queue/listen --verbose --isolated
```

```sh
php craft gc --isolated
```

**Reply by aaronbushnell** [reply]
Thanks, @brandonkelly! Just to make sure I'm following, is this the expected behavior of `--isolated`:

- You have a load-balanced environment with three servers
- Server 1 is running `queue/run --isolated`
- If you try to run `queue/run --isolated` on Server 2 it would fail because Server 1 is already running it?

**Reply by brandonkelly** [reply]
Correct. Assuming that

1. Dev Mode is disabled on both servers
2. Both servers are [sharing the same mutex source](https://craftcms.com/knowledge-base/configuring-load-balanced-environments#mutex-locks) (e.g. a Redis server)

**Reply by aaronbushnell** [reply]
That's awesome! Thanks a ton, @brandonkelly!

**Reply by brandonkelly** [reply]
4.4 is out!

**Comment by aaronbushnell** [comment]
Hey @brandonkelly — I noticed something about #12350. 

When you call...
```php
if (!Craft::$app->getMutex()->acquire($name)) {
    $this->stderr("The $uniqueId command is already running.\n", Console::FG_RED);
    return false;
}
```
...it will correctly check if the given command is running, but the Yii2 mutex-redis package has [a default TTL of 30 seconds](https://github.com/yiisoft/mutex-redis/blob/master/src/RedisMutex.php#L34-L36). If your queue is ran every minute via a cron, the `acquire()` check will pass and another queue worker will start because the mutex lock was released via the TTL setting.

I wonder if this feature needs to somehow alter the TTL so those locks can persist until the task has truly finished?

**Reply by brandonkelly** [reply]
Thanks for pointing that out! You actually linked to the Yii 3 Redis Mutex driver, but `yii2-redis` [has a similar `$expire` property](https://github.com/yiisoft/yii2-redis/blob/master/src/Mutex.php#L67).

We just discussed this internally and decided the best way to handle this is to recommend increasing `expire` to 900 seconds (15 minutes), at least for console requests.

The [`mutex` example in the docs](https://craftcms.com/docs/4.x/config/app.html#mutex) has been updated to reflect that, and I’ve also [added a warning](https://github.com/craftcms/cms/commit/faa0eb1c166b9bd2be9720ae909f32c47eeae4be) to the CLI output when using `--isolated`, if Redis’ `expire` is set to less than `900`:

<img width="639" alt="A warning output within a console command explaining that the mutex component should be configured with `expire` set to at least 900." src="https://user-images.githubusercontent.com/47792/217057019-2a6d12f6-542f-4a7e-956a-677e7fd4735a.png">

(Yes, some commands can go longer than 15 minutes currently, but that won’ be the case in 4.4.)

**Reply by aaronbushnell** [reply]
You bet, @brandonkelly! So, with a timeout of 15 minutes, that would mean—
- You run `php craft queue/run --isolated` 
- It holds a mutex lock to ensure other queue runners do not start
- The mutex lock is auto-released after 15 minutes
- If the queue is still running then the next scheduled `queue/run --isolated` command introduces a second worker

Is that accurate?

**Reply by brandonkelly** [reply]
@aaronbushnell Correct. But we are working to improve queue functionality a bit in Craft 4.4 to avoid super long running queues (see #12638 - WIP).

**Reply by aaronbushnell** [reply]
Awesome, thanks, @brandonkelly!
